packages/jest-cli/src/SearchSource.js

```diff
@@ -158,29 +158,48 @@ export default class SearchSource {
       buildSnapshotResolver(this._context.config),
     );
 
-    const tests = toTests(
-      this._context,
-      dependencyResolver.resolveInverse(
-        allPaths,
-        this.isTestFilePath.bind(this),
-        {
-          skipNodeResolution: this._context.config.skipNodeResolution,
-        },
-      ),
+    if (!collectCoverage) {
+      return {
+        tests: toTests(
+          this._context,
+          dependencyResolver.resolveInverse(
+            allPaths,
+            this.isTestFilePath.bind(this),
+            {skipNodeResolution: this._context.config.skipNodeResolution},
+          ),
+        ),
+      };
+    }
+
+    const testModulesMap = dependencyResolver.resolveInverseModuleMap(
+      allPaths,
+      this.isTestFilePath.bind(this),
+      {skipNodeResolution: this._context.config.skipNodeResolution},
     );
-    let collectCoverageFrom;
 
-    // If we are collecting coverage, also return collectCoverageFrom patterns
-    if (collectCoverage) {
-      collectCoverageFrom = Array.from(allPaths).map(filename => {
+    const allPathsAbsolute = Array.from(allPaths).map(p => path.resolve(p));
+
+    return {
+      collectCoverageFrom: Array.from(
+        testModulesMap.reduce((acc, testModule) => {
+          if (testModule.dependencies) {
+            testModule.dependencies
+              .filter(p => allPathsAbsolute.some(ap => ap === p))
+              .forEach(p => acc.add(p));
```

Comment:
Maybe it's just me, but I find mutating the accumulated value in `reduce` super weird. Can we maybe move all the processing out of `return` and just `forEach` over `testModulesMap`?

Reply:
Mutating the accumulated is a common use case of `reduce`, do you know of any drawbacks to it other than "finding it weird"?

I can create a variable `collectCoverageFrom` and then assign it to the result, but it would be an unnecessary variable to create as the values can be converted in a single expression.

Reply:
Is it? I've never seen it used that way.
The concept comes from functional programming, the normal way to use it is combining the `acc` and `next` into a new value and returning that.
I also think in general it wouldn't hurt to break up that enormous `return` statement?